test_rmw_implementation/test/test_wait_set.cpp

```diff
@@ -0,0 +1,322 @@
+    &timeout_argument);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription and guard_conditions with no timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, nullptr, nullptr, nullptr, wait_set,
+    &timeout_argument_zero);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription, guard_conditions and services with 100ms timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, nullptr, nullptr, wait_set,
+    &timeout_argument);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription, guard_conditions and services with no timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, nullptr, nullptr, wait_set,
+    &timeout_argument_zero);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription, guard_conditions, services and clients with 100ms timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, clients, nullptr, wait_set,
+    &timeout_argument);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription, guard_conditions, services and clients with no timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, clients, nullptr, wait_set,
+    &timeout_argument_zero);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription, guard_conditions, services, clients and events with 100ms
+  // timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, clients, events, wait_set,
+    &timeout_argument);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  // Used a valid wait_set, subscription, guard_conditions, services, clients and events with no
+  // timeout
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, clients, events, wait_set,
+    &timeout_argument_zero);
+  EXPECT_EQ(ret, RMW_RET_TIMEOUT) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  const char * implementation_identifier = wait_set->implementation_identifier;
+  wait_set->implementation_identifier = "rmw_baddss_cpp";
+  ret = rmw_wait(
+    subscriptions, guard_conditions, services, clients, events, wait_set,
+    &timeout_argument);
+  EXPECT_EQ(ret, RMW_RET_INCORRECT_RMW_IMPLEMENTATION) << rcutils_get_error_string().str;
+  rmw_reset_error();
+
+  wait_set->implementation_identifier = implementation_identifier;
+
+  // Battle test rmw_wait.
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    int64_t count = rcutils_fault_injection_get_count();
+    rcutils_fault_injection_set_count(RCUTILS_FAULT_INJECTION_FAIL_NOW);
```

Comment:
@ahcorde nit: I think you don't need to play with `count` in this case -- just let the fault injection test do the work.